engine/order_manager.go

```diff
@@ -1070,7 +1078,7 @@ func (s *store) add(det *order.Detail) error {
 	if err != nil {
 		return err
 	}
-	if s.exists(det) { // TODO: Error on conflict; remove unnecessary locking.
+	if s.exists(det) {
```

Comment:
Maybe reintroduce the `TODO: remove unnecessary RW locking.` There is a duplication potential just by looking at the code, if this function is called by different routines for the same order ID which is unlikely because the function is not exported but it highlights it might need a possible redesign.

Reply:
Okay, so to be clear:
We're about to Write lock it anyway, therefore you'd like to optimise not RLocking inside exists/getByDetails.
So you'd like to leave a TODO for that later.
I considered this. I'm pretty sure it's a premature optimisation, since add() is a write method.
I don't like TODOs that aren't coupled with actionable task numbers. They nearly never get done.
Suppose we did open an issue for this. I can't see it ever getting to the top of anyone's list.

TL;DR: It was a considered removal of the TODO because it's non-trivial, and premopt.

I'll readd it if you still want, though 😄

Reply:
It's not an optimisation per se it's more of a protection, when a RWMutex is not write locked the calling function doesn't semaquire. So just from a quick observation I may be wrong but concurrently if you call exists (read lock) and two routines for example could slip through and call write lock `s.m.Lock()` below which will act as a turnstile gate thus potentially producing a doubling of append calls. 

You are correct TODOs are not the best. 

Option 1) Re add TODO or Option 2) Fix the potential issue while you are there, which is remove the ReadLock and Unlock calls from `exists` and insert them into the scope of the calling function.

Reply:
Gotcha. You're right, I wasn't understanding your concern, which is definitely valid.

Whenever I feel I'm being annoyingly verbose it pays off.

Will Make-It-So.

Reply:
Tried Option 2 but it tasted awful because we're not *in* the store when we'd be fondling it's mutex.
I tried to live with it but I ended up rocking back and forth on the floor of the shower.
Instead I've opted to internalise it.
Is ee25cb74 acceptable?

Reply:
Yeah that is a way better approach thanks. 🚀  Thinking about it last night I don't think we should be having a RWmutex anyway, it puts a lot of burden on the writers. I will mull this over.